planner/core/plan_cache_lru.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/pingcap/tidb/util/hack"
+	"github.com/pingcap/tidb/util/kvcache"
+)
+
+// planCacheEntry wraps Key and Value. It's the value of list.Element.
+type planCacheEntry struct {
+	PlanKey   kvcache.Key
+	PlanValue kvcache.Value
+}
+
+// LRUPlanCache is a dedicated least recently used cache, JUST use for plan cache.
+type LRUPlanCache struct {
+	capacity uint
+	size     uint
+	// buckets replace the map in general LRU
+	buckets map[string]map[*list.Element]struct{}
+	lruList *list.List
+	// lock make cache thread safe
+	lock sync.Mutex
+
+	// pickFromBucket get one element from bucket. The LRUPlanCache can not work if it is nil
+	pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool)
+	// onEvict will be called if any eviction happened, only for test use now
+	onEvict func(kvcache.Key, kvcache.Value)
+}
+
+// NewLRUPlanCache creates a PCLRUCache object, whose capacity is "capacity".
+// NOTE: "capacity" should be a positive value.
+func NewLRUPlanCache(capacity uint,
+	pickFromBucket func(map[*list.Element]struct{}, []*types.FieldType) (*list.Element, bool)) (*LRUPlanCache, error) {
+	if capacity < 1 {
+		return nil, errors.New("capacity of LRU Cache should be at least 1")
+	}
+	return &LRUPlanCache{
+		capacity:       capacity,
+		size:           0,
+		buckets:        make(map[string]map[*list.Element]struct{}, 1), //Generally one query has one plan
+		lruList:        list.New(),
+		pickFromBucket: pickFromBucket,
+	}, nil
+}
+
+// Get tries to find the corresponding value according to the given key.
+func (l *LRUPlanCache) Get(key kvcache.Key, paramTypes []*types.FieldType) (value kvcache.Value, ok bool) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	bucket, bucketExist := l.buckets[string(hack.String(key.Hash()))]
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			l.lruList.MoveToFront(element)
+			return element.Value.(*planCacheEntry).PlanValue, true
+		}
+	}
+	return nil, false
+}
+
+// Put puts the (key, value) pair into the LRU Cache.
+func (l *LRUPlanCache) Put(key kvcache.Key, value kvcache.Value, paramTypes []*types.FieldType) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := string(key.Hash())
+	bucket, bucketExist := l.buckets[hash]
+	// bucket exist
+	if bucketExist {
+		if element, exist := l.pickFromBucket(bucket, paramTypes); exist {
+			element.Value.(*planCacheEntry).PlanValue = value
+			l.lruList.MoveToFront(element)
+			return
+		}
+	} else {
+		l.buckets[hash] = make(map[*list.Element]struct{}, 1)
+	}
+
+	newCacheEntry := &planCacheEntry{
+		PlanKey:   key,
+		PlanValue: value,
+	}
+	element := l.lruList.PushFront(newCacheEntry)
+	l.buckets[hash][element] = struct{}{}
+	l.size++
+	if l.size > l.capacity {
+		l.removeOldest()
+	}
+}
+
+// Delete deletes the multi-values from the LRU Cache.
+func (l *LRUPlanCache) Delete(key kvcache.Key) {
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	hash := hack.String(key.Hash())
```

Comment:
Why in the `put` function use the `hash := string(key.Hash())`, but here use the `hack.String()`?

Reply:
> Why in the `put` function use the `hash := string(key.Hash())`, but here use the `hack.String()`?

Here use hack.String to avoid unnecessary copy. But we need ensure the key in cache is not changeable, so convert it to `string` in `Put` function